test/Microsoft.Azure.Devices.Edge.Test.Common/LeafDevice.cs

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Test.Common
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Globalization;
+    using System.IO;
+    using System.Linq;
+    using System.Net;
+    using System.Security.Cryptography.X509Certificates;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Edge.Test.Common.Certs;
+    using Microsoft.Azure.Devices.Edge.Test.Common.Config;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Serilog;
+
+    public class LeafDevice
+    {
+        readonly DeviceClient client;
+        readonly Device device;
+        readonly IotHub iotHub;
+        readonly string messageId;
+
+        LeafDevice(Device device, DeviceClient client, IotHub iotHub)
+        {
+            this.client = client;
+            this.device = device;
+            this.iotHub = iotHub;
+            this.messageId = Guid.NewGuid().ToString();
+        }
+
+        public static Task<LeafDevice> CreateAsync(
+            string leafDeviceId,
+            Protocol protocol,
+            AuthenticationType auth,
+            Option<string> parentId,
+            bool useSecondaryCertificate,
+            EdgeCertificateAuthority edgeCa,
+            IotHub iotHub,
+            CancellationToken token)
+        {
+            return Profiler.Run(
+                async () =>
+                {
+                    ITransportSettings transport = protocol.ToTransportSettings();
+                    Platform.InstallEdgeCertificates(edgeCa.Certificates.TrustedCertificates, transport);
+
+                    string edgeHostname = Dns.GetHostName().ToLower();
```

Comment:
should it be passing in? It looks weird to me as leaf device may not always be run on the same host as edge device does.

Reply:
Currently the leaf and edge devices are always on the same host. We could move this call up into the test, but then that's just more work for the test to do, with no real benefit. I'd rather move it up if/when the architecture changes (when leaf devices can run from physically separate hosts).